### PR TITLE
Fixes #1257 Cloudways Varnish Purge auto-compatibility

### LINKED ISSUE #1257: Cloudways - Varnish cache is not cleared when the Varnish add-on is enabled
**Describe the bug**
On Cloudways, Varnish cache is not purged when WP Rocket's cache is cleared (Varnish add-on is enabled). 

This leads to 404 on minified files.

**Steps to reproduce the issue**
1. Use a Cloudways-hosted website where Varnish is enabled.  We don't have one currently, so we should create it from scratch.
2. Enable the Varnish add-on.
3. Make a change to a post.
4. Clear WP Rocket's cache.
5. Check, in an incognito browser window, for the updated post.

**Related Ticket**
https://secure.helpscout.net/conversation/686147105/83005/?folderId=2135277

### PR DESCRIPTION
**Pull request by remyperona**
Fixes #1257 

- [x] Create subscriber for Cloudways compatibility
- [x] Write tests
- [x] QA

### CODE REVIEW

File: inc/ThirdParty/Hostings/Cloudways.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	/**
+	 * Returns true
+	 *
+	 * @since 3.5.5
+	 *
+	 * @return true
+	 */
+	public function return_true() {
+		return true;
+	}
```
Comment (hellofromtonya) [inline inc/ThirdParty/Hostings/Cloudways.php:53]: Why not return `__return_true` and `__return_false`?

File: inc/ThirdParty/Hostings/Cloudways.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	 * @since 3.5.5
+	 *
+	 * @param array $varnish_ip Varnish IP.
+	 * @return array
+	 */
+	public function varnish_ip( $varnish_ip ) {
+		$varnish_ip[] = '127.0.0.1:8080';
+
+		return $varnish_ip;
+	}
```
Comment (hellofromtonya) [inline inc/ThirdParty/Hostings/Cloudways.php:85]: Will the value passed into this callback always be an array?

- If yes, type hint it.
- If no, then first initialize it as an array if it's not one.

Add a test to validate it.

File: inc/ThirdParty/Hostings/Cloudways.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	 */
+	public function varnish_addon_title( $settings ) {
+		$settings['varnish_auto_purge']['title'] = sprintf(
+			// Translators: %s = Hosting name.
+			__( 'Your site is hosted on %s, we have enabled Varnish auto-purge for compatibility.', 'rocket' ),
+			'Cloudways'
+		);
+
+		return $settings;
+	}
```
Comment (hellofromtonya) [inline inc/ThirdParty/Hostings/Cloudways.php:71]: Will the value passed into this callback always be an array?

- If yes, type hint it.
- If no, then first initialize it as an array if it's not one and/or add the first key and set to `[]` if it doesn't exist.

Add a test to validate it.

File: tests/Integration/inc/ThirdParty/Hostings/Cloudways/varnishIP.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * @covers \WP_Rocket\ThirdParty\Hostings\Cloudways::varnish_ip
+ * @group Cloudways
+ * @group ThirdParty
+ */
+class Test_VarnishIP extends TestCase {
+	public function testShouldReturnCloudwaysVarnishIP() {
+		$varnish_ip = [];
+		$expected   = [
+			'127.0.0.1:8080',
+		];
```
Comment (hellofromtonya) [inline tests/Integration/inc/ThirdParty/Hostings/Cloudways/varnishIP.php:17]: Move to test data fixture. Add more use cases such as:

- An empty array is given
- Other IPs are given
- If a non-array _could_ be given, add that

File: tests/Integration/inc/ThirdParty/Hostings/Cloudways/varnishAddonTitle.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+			'varnish_auto_purge' => [
+				'title' => '',
+			],
+		];
+
+		$expected = [
+			'varnish_auto_purge' => [
+				'title' => 'Your site is hosted on Cloudways, we have enabled Varnish auto-purge for compatibility.'
+			],
+		];
```
Comment (hellofromtonya) [inline tests/Integration/inc/ThirdParty/Hostings/Cloudways/varnishAddonTitle.php:24]: Move to test data fixture. Add more use cases such as:

- An empty settings array is given
- Other settings are given
- If a non-array _could_ be given, pass `false` or `null`.

File: tests/Unit/inc/ThirdParty/Hostings/Cloudways/varnishAddonTitle.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+				'title' => '',
+			],
+		];
+
+		$expected = [
+			'varnish_auto_purge' => [
+				'title' => 'Your site is hosted on Cloudways, we have enabled Varnish auto-purge for compatibility.'
+			],
+		];
+
```
Comment (hellofromtonya) [inline tests/Unit/inc/ThirdParty/Hostings/Cloudways/varnishAddonTitle.php:29]: Same here. Use the test data fixture.

File: tests/Unit/inc/ThirdParty/Hostings/Cloudways/varnishIP.php [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * @group ThirdParty
+ */
+class Test_VarnishIP extends TestCase {
+	public function testShouldReturnCloudwaysVarnishIP() {
+		$cloudways = new Cloudways();
+
+		$varnish_ip = [];
+		$expected   = [
+			'127.0.0.1:8080',
+		];
```
Comment (hellofromtonya) [inline tests/Unit/inc/ThirdParty/Hostings/Cloudways/varnishIP.php:19]: Same here. Use the test data fixture.

**Review by hellofromtonya — CHANGES_REQUESTED** [review]

File: inc/ThirdParty/Hostings/Cloudways.php (hunk 1 above)
Comment (Tabrisrp) [reply inc/ThirdParty/Hostings/Cloudways.php:53]: With the subscriber system currently, we can't use a global callback like those functions in `get_subscribed_events()`, it needs to be a method of the class.

Indeed I would have used those instead if possible to not re-invent the wheel.

**Review by hellofromtonya — APPROVED** [review]

**Review by crystinutzaa — APPROVED** [review]

**Comment by hellofromtonya** [comment]
@wp-media/dev-qa This PR is ready for QA.

**Comment by piotrbak** [comment]
Starting the QA

**Review by piotrbak — APPROVED** [review]
It worked like expected on a test site on Cloudways, I also tested it with Cloudflare enabled, all good!

**Pull request merged**